proof-systems/src/darlin/poly-commit/src/constraints/mod.rs

```diff
@@ -0,0 +1,447 @@
+        NonNativeFieldGadget::<G::ScalarField, ConstraintF>::from_bits(
+            cs,
+            challenge
+                .iter()
+                .rev()
+                .cloned()
+                .collect::<Vec<_>>()
+                .as_slice(),
+        )
+    }
+
+    /// Succinct check of the verify
+    fn succinct_verify<CS: ConstraintSystemAbstract<ConstraintF>>(
+        cs: CS,
+        vk: &Self::VerifierKey,
+        commitment: &Self::Commitment,
+        point: &NonNativeFieldGadget<G::ScalarField, ConstraintF>,
+        value: &Vec<Boolean>, // bits of evaluation point in big-endian order
+        proof: &Self::Proof,
+        random_oracle: &mut Self::RandomOracle,
+    ) -> Result<Self::VerifierState, Self::Error>;
+
+    /// succinct check of the verification of an opening proof for multiple polynomials in the
+    /// same point
+    fn succinct_verify_single_point_multi_poly<CS: ConstraintSystemAbstract<ConstraintF>>(
+        mut cs: CS,
+        vk: &Self::VerifierKey,
+        labeled_commitments: &[LabeledCommitmentGadget<Self, ConstraintF, G, PC>], //ToDo: see if we can take an impl IntoIterator as for the primitive
+        point: &NonNativeFieldGadget<G::ScalarField, ConstraintF>,
+        values: &[NonNativeFieldGadget<G::ScalarField, ConstraintF>],
+        proof: &Self::Proof,
+        random_oracle: &mut Self::RandomOracle,
+    ) -> Result<Self::VerifierState, Self::Error> {
+        let lambda = random_oracle.enforce_get_challenge::<_, 128>(
+            cs.ns(|| "squeeze lambda for single-point-multi-poly verify"),
+            )?;
+        let lambda_non_native = Self::challenge_to_non_native_field_element(
+            cs.ns(|| "convert lambda to non native field gadget"),
+            &lambda,
+        )?;
+        let default_commitment = LabeledCommitmentGadget::new(
+            String::from("default commitment"),
+            Self::Commitment::zero(&mut cs)?,
+        );
+        let mut commitments_iter = labeled_commitments.iter().rev();
+        let mut batched_commitment = commitments_iter
+            .next()
+            .unwrap_or(&default_commitment)
+            .commitment()
+            .clone();
+
+        let zero_value = NonNativeFieldGadget::<G::ScalarField, ConstraintF>::zero(&mut cs)?;
+        let mut values_iter = values.iter().rev();
+        let mut batched_value = values_iter.next().unwrap_or(&zero_value).clone();
+
+        for (i, (commitment, value)) in commitments_iter.zip(values_iter).enumerate() {
+            batched_commitment = Self::mul_by_challenge(
+                cs.ns(|| format!("lambda*batched_commitment_{}", i)),
+                &batched_commitment,
+                lambda.iter(),
+            )?;
+            batched_commitment = batched_commitment.add(
+                cs.ns(|| format!("add commitment {} to batched_commitment", i)),
+                commitment.commitment(),
```

Comment:
Couple of comments here:

- If we assume commitments can't be the infinity point (I don't think it can ever happen with our commit algorithm), and lambda is not zero either, then using safe_mul it's useless;
- For the same reason, unless we assume two commitments are the same (and I think it cannot happen either), using complete add is unneeded too.

We can save some constraints. Exceptional cases cannot happen by construction. Can still happen with negligible probability but we can safely ignore it

Reply:
This is true everytime you batch, so I won't bother commenting all the instances. Please check the code and be sure that you use safe_mul_bits and complete_add only when exceptional cases can actually happen by construction. Ignore, instead, all the cases where this can happen with negligible probability